### PR TITLE
Line opacity parallelization

### PR DESCRIPTION
**Pull request by jvshields**
This expands PR #178 to allow for parallelization of the slowest part of the code. It should not be merged before that pull request.

### CODE REVIEW

**Review by DeerWhale — APPROVED** [review]

File: stardis/radiation_field/opacities/opacities_solvers/base.py [hunk 1]
```diff
@@ -448,29 +443,116 @@ def calc_alpha_line_at_nu(
...
+def calc_alan_entries_parallel(
+    no_of_depth_points,
+    tracing_nus_values,
+    delta_nus,
+    doppler_widths,
+    gammas,
+    alphas_array,
+    broadening_range=None,
+    h_lines_indices=None,
+):
```
Comment (andrewfullard) [inline stardis/radiation_field/opacities/opacities_solvers/base.py:484]: Docstring missing

File: stardis/radiation_field/opacities/opacities_solvers/base.py [hunk 2]
```diff
@@ -448,29 +443,116 @@ def calc_alpha_line_at_nu(
...
+    alpha_line_at_nu = np.zeros((no_of_depth_points, len(tracing_nus_values)))
+
+    if broadening_range is None:
+        for frequency_index in numba.prange(len(tracing_nus_values)):
```
Comment (andrewfullard) [inline stardis/radiation_field/opacities/opacities_solvers/base.py:488]: Is there any configuration of the number of threads allowed for Numba?

File: stardis/radiation_field/opacities/opacities_solvers/base.py [hunk 3]
```diff
@@ -448,29 +443,116 @@ def calc_alpha_line_at_nu(
...
 def calc_alan_entries(
+    no_of_depth_points,
+    tracing_nus_values,
+    delta_nus,
+    doppler_widths,
+    gammas,
+    alphas_array,
+    broadening_range=None,
+    h_lines_indices=None,
+):
```
Comment (andrewfullard) [inline stardis/radiation_field/opacities/opacities_solvers/base.py:524]: Missing docstring

**Review by andrewfullard — CHANGES_REQUESTED** [review]

File: stardis/radiation_field/opacities/opacities_solvers/base.py (hunk 2 above)
Comment (jvshields) [reply stardis/radiation_field/opacities/opacities_solvers/base.py:488]: Yup. You can do numba.set_num_threads(int) to tell numba how many it is allowed to use. I think by default it tries to use all of them, but that's not guaranteed and you can see how many with numba.config.NUMBA_NUM_THREADS.

Comment (jvshields) [reply stardis/radiation_field/opacities/opacities_solvers/base.py:519]: I was thinking about letting the user set the max number of threads in the config, but I don't know if that's better than just letting them interact with numba directly. If we stick with the latter case, then we can probably just detail how to do that in a parallelization tutorial notebook that could be thrown together.

Comment (andrewfullard) [reply stardis/radiation_field/opacities/opacities_solvers/base.py:519]: I was thinking something user-configurable like TARDIS. Depends entirely on how accessible you want parallelization to be.

Comment (jvshields) [reply stardis/radiation_field/opacities/opacities_solvers/base.py:519]: That makes sense and I've updated it to behave that way. You specify a number of threads in the config. -99 (default) is serial, 0 is max number of threads, and any positive integer will be passed to numba. There's also a logger.info message stating what happens based on that number.

Comment (jvshields) [reply stardis/radiation_field/opacities/opacities_solvers/base.py:519]: I changed this to 1 for serial, and -99 for max threads

**Review by DeerWhale — APPROVED** [review]

**Review by andrewfullard — APPROVED** [review]
Looks great!

**Pull request merged**